conftest.py

```diff
@@ -453,6 +281,280 @@ def tracy_profile():
+
+    # store test results for each phase of a call, which can
+    # be "setup", "call", "teardown"
+    item.stash.setdefault(phase_report_key, {})[rep.when] = rep
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_runtest_teardown(item, nextitem):
+    yield
+    metal_timeout_enabled = item.config.getoption("--metal-timeout")
+    using_xdist = int(os.getenv("PYTEST_XDIST_WORKER_COUNT", "0"))
+
+    if metal_timeout_enabled is not None or using_xdist:
+        report = item.stash[phase_report_key]
+        test_failed = report.get("call", None) and report["call"].failed
+        if test_failed:
+            logger.info(f"In custom teardown, open device ids: {set(item.pci_ids)}")
+            reset_tensix(set(item.pci_ids))
+
+
+# This is overriding the timer setup hook from pytest-timeout
+# If --metal-timeout is passed, we define a new timeout method that spawns a timer process
+# At timeout, the process kills it's parent (the test process) and then itself
+@pytest.hookimpl(tryfirst=True)
+def pytest_timeout_set_timer(item, settings):
+    metal_timeout_enabled = item.config.getoption("--metal-timeout")
+    using_xdist = int(os.getenv("PYTEST_XDIST_WORKER_COUNT", "0"))
+
+    if metal_timeout_enabled is not None or using_xdist:
+        parent_pid = os.getpid()
+        logger.info(f"Metal timeout {settings.timeout} seconds {parent_pid} for {item.nodeid}")
+
+        def get_parent_status():
+            try:
+                parent = psutil.Process(parent_pid)
+            except:
+                return "already dead"
+            return parent.status()
+
+        def run_timer(settings):
+            logger.info(f"Timer started for {item.nodeid}")
+            dead_status = ["zombie", "dead", "already dead"]
+            timeout = settings.timeout
+            parent_status = "running"
+            while parent_status not in dead_status and timeout > 0:
+                time.sleep(5)
+                timeout -= 5
+                parent_status = get_parent_status()
+            if parent_status != "already dead":
+                logger.info(f"Timing out test case")
+                os.kill(parent_pid, signal.SIGKILL)
+            logger.info(f"Killing timer")
+            os._exit(1)
+
+        def cancel():
+            logger.info(f"Cancelling timer")
+            metal_timer.terminate()
+
+        metal_timer = multiprocess.Process(target=run_timer, args=(settings,), daemon=True)
+        item.cancel_timeout = cancel
+        metal_timer.start()
+        # logger.info(f"parent and metal timer pid: {parent_pid} {metal_timer.pid}")
+    return True
+
+
+# This is a hook used in pytest-xdist to handle when a worker crashes out
+# In our case, combined with pytest-timeout thread method, the worker will crash out for a hang and
+# then it should get cleaned up by the controller through this fixture :fingers_crossed:
+@pytest.hookimpl(tryfirst=True)
+def pytest_handlecrashitem(crashitem, report, sched):
+    reset_tensix()
```

Comment:
Did you run post commit on this yet? What happens on cpu-only runners on GitHub with no CI reset scripts which have no TT stuff installed? Should we put in extra handling to check for the reset scripts explicitly in the reset_tensix fn?

Reply:
Post-commit will not use the xdist+timeout infra since tests shouldn't. Giving it a run anyways just to be sure

Reply:
I thought the plan was to have all tests auto use this?

Reply:
For now, I believe we mainly wanted this for pipelines that aren't being monitored as often/closely and take longer time. 

T3k pipelines are great candidates since they take so long and struggle to run all the way through.

Reply:
ok sounds great to me